### DISCUSSION TITLE
Is there an API to get the Device ID or MAC Address

### DISCUSSION THREAD

**Discussion by ahmed-abdelrazek** [unoplatform/uno#10156] [Q&A] [answered]
is there a way to get the device unique identifier or its mac address?
the documentation isn't clear about that 
https://platform.uno/docs/articles/features/windows-networking.html
https://platform.uno/docs/articles/features/windows-system-profile.html

**Comment by pkar70** [comment]
"get device MAC address" is something like oxymoron. You cannot get such address, as device hasn't such address. Network interface has. And device have many network interfaces: WiFi, Bluetooth, WAN (cellular), Ethernet in laptop or desktop, Ethernet in docking station, or in USB-C hub...

**Reply by ahmed-abdelrazek** [reply]
so let me be more precise the mac address for whatever network interface the user is using to connect to the internet at the moment or at least the device wi-fi mac address from the status page in about device in settings (android)

i just want someway to tell the api that this device is this device to enforce licensing even if i reinstalled the app but i'm heading more and more toward generating my own unique id at the app first launch

**Comment by pkar70** [comment]
Ok, so you want MAC address of one of these network interfaces that can be used to access Internet :)
Device can have multiple interfaces that can be used, simultanously, to access Internet; IP (or IP routing) decides which interface should be used to send/receive data to/from given host. E.g. for github.com it can be WiFi, for platform.uno - Ethernet, and for microsoft.com - Bluetooth :)

Yes, you have to use your own ID. Simply because e.g. Windows can create random MAC address for every connection by WiFi («To enable MAC randomization in Windows 10, you need to do the following. Go to Network & Internet -> WiFi. Under the WiFi adapter for which you wish to enable this, click the Advanced options link: Turn on the option called "Random Hardware Addresses" and you are done: Set it to "on" or "change daily"»)

**Reply by ahmed-abdelrazek** [reply]
i'm aware but all of that still doesn't answer my question does uno offer an api that can get me this info?!

**Comment by jeromelaban** [comment, marked as answer]
Most devices treat device identification as a privacy invading API, which means that it may be obfuscated or changed between reinstallations, depending on what the user asks. Uno does not implement [GetSystemIdForPublisher](https://learn.microsoft.com/en-us/uwp/api/windows.system.profile.systemidentification.getsystemidforpublisher?view=winrt-18362#remarks), though, which could be used to implement such a feature on platforms supporting a similar feature. If you're up for it, we're looking for contributors.

Depending on platforms, you can use [native APIs](https://learn.microsoft.com/en-us/dotnet/api/uikit.uidevice.identifierforvendor?view=xamarin-ios-sdk-12) to get advertisement identifiers (iOS/Android), but it won't be a good way to lock down your app for licensing, in the same way using the MAC address can be changed by the user.

If you aim at mobile platforms (iOS/Android), it would be best using the platform's Store APIs for in-app purchases. For the other platforms, like WebAssembly, it's somehow more difficult or easily circumventable.

**Reply by ahmed-abdelrazek** [reply]
unfortunately i don't have any experience with android ios or mac development to implement such feature that's why i turned to uno in the first place

i will be looking into store api for inapp purchases and thanks for the tip 
we will not use the ID or MAC to enforce the license on the device rather to check on the server if the admin/IT put that device on the allowed list to gain full access or just read only so if you didn't send one like in case of webassembly or the one you send isn't in the allowed list you will be limited
